File: MP4/src/main/java/edu/illinois/storm/FileReaderSpout.java
package edu.illinois.storm;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.Map;
import org.apache.storm.spout.SpoutOutputCollector;
import org.apache.storm.task.TopologyContext;
import org.apache.storm.topology.IRichSpout;
import org.apache.storm.topology.OutputFieldsDeclarer;
import org.apache.storm.tuple.Fields;
import org.apache.storm.tuple.Values;
import org.apache.storm.utils.Utils;

/** a spout that generate sentences from a file */
public class FileReaderSpout implements IRichSpout {
  private SpoutOutputCollector _collector;
  private TopologyContext _context;
  private String inputFile;

  // Hint: Add necessary instance variables if needed
  private BufferedReader reader;

  @Override
  public void open(Map conf, TopologyContext context, SpoutOutputCollector collector) {
    this._context = context;
    this._collector = collector;

    /* ----------------------TODO-----------------------
    Task: initialize the file reader
    ------------------------------------------------- */
    try {
      reader = new BufferedReader(new FileReader(inputFile));
    } catch (FileNotFoundException e) {
        e.printStackTrace();
    }
    // END

  }

  // Set input file path
  // public FileReaderSpout withInputFileProperties(String inputFile) {
  //   this.inputFile = inputFile;
  //   return this;
  // }
  public void setInputFile(String inputFile) {
    this.inputFile = inputFile;
  }

  @Override
  public void nextTuple() {

    /* ----------------------TODO-----------------------
    Task:
    1. read the next line and emit a tuple for it
    2. don't forget to add a small sleep when the file is entirely read to prevent a busy-loop
    ------------------------------------------------- */
    try{
      while(reader.ready()) {
          String line = reader.readLine();
          _collector.emit(new Values(line));
      }
    } catch(IOException e){
        e.printStackTrace();
    }
    Utils.sleep(1000);
    // END
  }

  @Override
  public void declareOutputFields(OutputFieldsDeclarer declarer) {
    /* ----------------------TODO-----------------------
    Task: define the declarer
    ------------------------------------------------- */
    declarer.declare(new Fields("word"));
    // END
  }

  @Override
  public void close() {
    /* ----------------------TODO-----------------------
    Task: close the file
    ------------------------------------------------- */
    try {
      reader.close();
    } catch (IOException e) {
      e.printStackTrace();
    }
    // END

  }

  public void fail(Object msgId) {}

  public void ack(Object msgId) {}

  @Override
  public void activate() {}

  @Override
  public void deactivate() {}

  @Override
  public Map<String, Object> getComponentConfiguration() {
    return null;
  }
}

File: MP4_python/multilang/resources/file_reader_spout.py
# import os
# from os.path import join
from time import sleep

# from streamparse import Spout
import storm

class FileReaderSpout(storm.Spout):

    def initialize(self, conf, context):
        self._conf = conf
        self._context = context
        self._complete = False

        storm.logInfo("Spout instance starting...")

        # TODO:
        # Task: Initialize the file reader
        fname = "/tmp/data.txt"
        self.fr = open(fname, "r") # add file reader here
        # End

    def nextTuple(self):
        # TODO:
        # Task 1: read the next line and emit a tuple for it
        for line in self.fr.readlines():
            # Emit a random sentence
            # storm.logInfo("Emiting %s" % line)
            line = line.strip()
            storm.emit([line])

        # Task 2: don't forget to sleep for 1 second when the file is entirely read to prevent a busy-loop
        sleep(1)
        self._complete = True
        # End
        


# Start the spout when it's invoked
FileReaderSpout().run()
